core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,874 @@
+   */
+  final def drop(n: Int): Chunk[A] =
+    if (n <= 0) self
+    else
+      self match {
+        case Chunk.Slice(c, o, l)        => Chunk.Slice(c, o + n, l - n)
+        case Chunk.Singleton(_) if n > 0 => Chunk.empty
+        case c @ Chunk.Singleton(_)      => c
+        case Chunk.Empty                 => Chunk.empty
+        case _                           => Chunk.Slice(self, n, self.length - n)
+      }
+
+  /**
+   * Drops all elements so long as the predicate returns true.
+   */
+  def dropWhile(f: A => Boolean): Chunk[A] = {
+    val len = self.length
+
+    var i = 0
+    while (i < len && f(self(i))) {
+      i += 1
+    }
+
+    if (i == len) Chunk.empty
+    else if (i == 0) this
+    else Chunk.Slice(self, i, len - i)
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Chunk[_] =>
+      if (self.length != that.length) false
+      else {
+        var i     = 0
+        var equal = true
+        val len   = self.length
+
+        while (equal && i < len) {
+          equal = self(i) == that(i)
+          i += 1
+        }
+
+        equal
+      }
+    case _ => false
+  }
+
+  /**
+   * Returns a filtered subset of this chunk.
+   */
+  def filter(f: A => Boolean): Chunk[A] = {
+    implicit val B: ClassTag[A] = Chunk.classTagOf(this)
+
+    val len  = self.length
+    val dest = Array.ofDim[A](len)
+
+    var i = 0
+    var j = 0
+    while (i < len) {
+      val elem = self(i)
+
+      if (f(elem)) {
+        dest(j) = elem
+        j += 1
+      }
+
+      i += 1
+    }
+
+    if (j == 0) Chunk.Empty
+    else Chunk.Slice(Chunk.Arr(dest), 0, j)
+  }
+
+  /**
+   * Returns the concatenation of mapping every element into a new chunk using
+   * the specified function.
+   */
+  def flatMap[@specialized B](f: A => Chunk[B]): Chunk[B] = {
+    val len                    = self.length
+    var chunks: List[Chunk[B]] = Nil
+
+    var i               = 0
+    var total           = 0
+    var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+    while (i < len) {
+      val chunk = f(self(i))
+      if (chunk.length > 0) B0 = Chunk.classTagOf(chunk)
+      chunks ::= chunk
+      i += 1
+      total += chunk.length
+    }
+    chunks = chunks.reverse
+
+    if (B0 == null) Chunk.empty
+    else {
+      implicit val B: ClassTag[B] = B0
+
+      val dest: Array[B] = Array.ofDim(total)
+
+      val it = chunks.iterator
+      var n  = 0
+      while (it.hasNext) {
+        val chunk = it.next
+
+        chunk.toArray(n, dest)
+        n += chunk.length
+      }
+
+      Chunk.Arr(dest)
+    }
+  }
+
+  final def foldMLazy[E, S](z: S)(pred: S => Boolean)(f: (S, A) => IO[E, S]): IO[E, S] = {
+    val len = length
+
+    def loop(s: S, i: Int): IO[E, S] =
+      if (i >= len) IO.now(s)
+      else {
+        if (pred(s)) f(s, self(i)).flatMap(loop(_, i + 1))
+        else IO.now(s)
+      }
+
+    loop(z, 0)
+  }
+
+  final def foldLeftLazy[S](z: S)(pred: S => Boolean)(f: (S, A) => S): S = {
```

Comment:
What is lazy in this function? 🤔

Reply:
Lol. I think we'd be open to a better name (cc @jdegoes). It's lazy in that it is short-circuiting and will stop traversing the chunk when `pred` returns `false`.